src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNISslStream.cs

```diff
@@ -0,0 +1,44 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Net.Security;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Data.SqlClient.SNI
+{
+    /// <summary>
+    /// This class extends SslStream to customize stream behavior for Managed SNI implementation.
+    /// </summary>
+    internal class SNISslStream : SslStream
+    {
+        private readonly ConcurrentQueueSemaphore _writeQueueSemaphore;
+        private readonly ConcurrentQueueSemaphore _readQueueSemaphore;
+
+        public SNISslStream(Stream innerStream, bool leaveInnerStreamOpen, RemoteCertificateValidationCallback userCertificateValidationCallback)
+            : base(innerStream, leaveInnerStreamOpen, userCertificateValidationCallback)
+        {
+            _writeQueueSemaphore = new ConcurrentQueueSemaphore(1);
+            _readQueueSemaphore = new ConcurrentQueueSemaphore(1);
+        }
+
+        // Prevent the ReadAsync's collision by running task in Semaphore Slim
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            _readQueueSemaphore.Wait();
+            Task<int> t = base.ReadAsync(buffer, offset, count, cancellationToken);
+            _readQueueSemaphore.Release();
```

Comment:
Shouldn't this be in a `.ContinueWith()` as used in `WriteAsync()`?

Reply:
I was trying something like this, but it made ReadAsync go synchronous:
```
public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
{
    _readQueueSemaphore.Wait();
    return base.ReadAsync(buffer, offset, count, cancellationToken)
        .ContinueWith(_ => { _readQueueSemaphore.Release(); return _.GetAwaiter().GetResult(); });
}
```

`Release()` is sync call, and returns `void`. Any ideas?

Reply:
Calling the sync Release in ReadAsync is async-over-sync, meaning that the thread will block instead being truly async, which is what ReadAsync advertises itself to be. If this is a long wait, this can cause thread pool starvation, as TP threads get synchronously blocked.

Is there any reason not to make this method async, and await on _readQueueSemaphore.WaitAsync?

Reply:
@cheenamalhotra I was suggesting that the pattern used in `WriteAsync()` should be used here as well by moving the `_readQueueSemaphore.Release()` call to ` .ContinueWith()` (minus all the `.GetAwaiter().GetResult()` stuff) - but in light of @roji 's comments, I guess neither pattern is correct.

Reply:
Looking at this again, I accidentally wrote "Release" where I meant to write "Wait" above. Release doesn't actually block, but Wait certainly does

Reply:
In that case, my original suggestion seems valid -- certainly the `.Wait()` will block, but that's exactly what has to happen to ensure that packets are read correctly.

```
            _readQueueSemaphore.Wait();
            return base.ReadAsync(buffer, offset, count, cancellationToken)
                .ContinueWith(t => _readQueueSemaphore.Release());
```

Reply:
@roji 

Also this is call to ConcurrentQueueSemaphore > `Wait()`, that delegates to Semaphore's `WaitAsync()` internally and waits on it. We can't call it async way as we need to block here.

Reply:
Thanks for the tip @roji !
I tested and hit perf issue with Blocking call to wait().. using WaitAsync() solves that issue, alongwith issue #422 is now getting fixed when specifying "Encrypt=true" ❤️ 

Thanks so much! :)